# Анализ и обработка результатов

В анализе данных важно наглядное представление результатов, чтобы их мог оценить заказчик. Это бизнес, где на кону громадный трафик, серверные мощности, личное время многотысячной аудитории. И одновременно это исследование.

Как в настоящей экспериментальной науке, ответ непредсказуем. Его точность зависит от умелого владения статистическими методами и от качества исходных данных. Перед тем, как начинать считать, проверьте, грамотно ли ваши данные подготовлены.

# Группировка данных

Слово «анализ» означает разбор, рассмотрение с разных сторон. Анализ данных начинают с разделения их на группы по какому-нибудь признаку. Эта операция называется группировка данных. Она помогает изучить материал более подробно, чтобы затем перейти к поиску взаимосвязей между отдельными группами.


Группировка оправданна, если данные чётко делятся по значимому признаку, а полученные группы близки к теме задачи. Например, когда есть данные обо всех покупках в супермаркете, можно смело заниматься группировкой. Так можно установить время наплыва покупателей и решить проблему пиковых нагрузок. Или посчитать средний чек — обычно для магазинов это ключевая метрика.


Стадии группировки хорошо укладываются в словесную формулу **split-apply-combine**:

•	разделить, **split** — разбиение на группы по определённому критерию;

•	применить, **apply** — применение какого-либо метода к каждой группе в отдельности, например, подсчёт численности группы методом count() или суммирование вызовом sum();

•	объединить, **combine** — сведение результатов в новую структуру данных, в зависимости от условий разделения и выполнения метода это бывает DataFrame и Series.

На картинке изображен принцип **split-apply-combine** для таблицы с экзопланетами. Посмотрим, как вообще идут дела с поиском экзопланет. Сначала данные делят по группам, где каждая группа — это год. Потом метод **count()** подсчитывает численность каждой группы. В итоге получаем новую структуру данных с группами, где каждая содержит год и число открытых за этот год экзопланет.

![image.png](attachment:image.png)

В Рandas для группировки данных есть метод **groupby()**. Он принимает как аргумент название столбца, по которому нужно группировать. В случае с делением экзопланет по годам открытия:

**print(exoplanet.groupby('discovered'))**

**<pandas.core.groupby.DataFrameGroupBy object at 0x7fc1e1ca3400>**


Применение метода **groupby()** к объекту типа DataFrame приводит к созданию объекта особого типа — **DataFrameGroupBy**. Это сгруппированные данные. Если применить к ним какой-нибудь метод Pandas, они станут новой структурой данных типа **DataFrame** или **Series**.
Подсчитаем сгруппированные по годам экзопланеты методом **count()**:

Если нужно сравнить наблюдения по одному показателю, метод применяют к **DataFrameGroupBy** с указанием на один столбец. Нас в первую очередь интересует радиус экзопланет: мы ищем другую Землю. Давайте получим таблицу с единственным столбцом 'radius':

**exo_number = exoplanet.groupby('discovered')['radius'].count()**

**print(exo_number)** 

Получили Series, где по годам открытия расписано количество экзопланет, для которых удалось установить радиус.

Посмотрим, как меняется средний радиус открытых экзопланет год от года. Для этого надо сложить радиусы планет, открытых за определённый год, и поделить на их количество (которое мы уже нашли).

Сумма радиусов считается методом **sum()**:

**exo_radius_sum = exoplanet.groupby('discovered')['radius'].sum()**


**print(exo_radius_sum)**



## Задача
Вашему вниманию предлагается датасет с обзором показателей самоубийств с 1985 по 2016 год — Сравнение социально-экономической информации с показателями самоубийств по годам и странам.

Столбцами этой таблицы являются: страна, год, пол, возрастная группа, количество самоубийств, население, уровень самоубийств, сводный ключ по стране за год, ИЧР (индекс человеческого развития) за год, ВВП за год, ВВП на душу населения, поколение (на основе среднего по возрастной группе).

1. Загрузите таблицу master, содержащую данные о самоубийствах. Выведите на экран таблицу и оцените данные, обратите внимание на названия столбцов, переименуйте их, если это необходимо.

2. Убедитесь в том, что данные прошли предподготовку.  Пропущенные и неопределённые значения выявляет метод isna(), а суммарное количество таких значений — метод sum().

3. Сгруппируйте DataFrame по столбцу country, сохраните полученный результат в переменной country_grouping.

4. Посчитайте количество случаев самоубийств по странам, которые были совершены, методом **sum()**, указав, что выбираем один столбец suicides_no.

Сохраните результат в переменной country_sum и выведите первые 30 строк этой таблицы.



In [14]:
#Загрузите таблицу master, содержащую данные о самоубийствах. 
#Выведите на экран таблицу и оцените данные, обратите внимание на названия столбцов, переименуйте их, если это необходимо.
import pandas as pd
df = pd.read_csv('master.csv')
df.head(5)



country    year     sex          age  suicides_no  population  \
0  Albania  1987.0    male  15-24 years         21.0    312900.0   
1  Albania  1987.0    male  35-54 years         16.0    308000.0   
2  Albania  1987.0  female  15-24 years         14.0    289700.0   
3  Albania  1987.0    male    75+ years          1.0     21800.0   
4  Albania  1987.0    male  25-34 years          9.0    274300.0   

   suicides/100k pop country-year  HDI for year  gdp_for_year ($)   \
0               6.71  Albania1987           NaN      2,156,624,900   
1               5.19  Albania1987           NaN      2,156,624,900   
2               4.83  Albania1987           NaN      2,156,624,900   
3               4.59  Albania1987           NaN      2,156,624,900   
4               3.28  Albania1987           NaN      2,156,624,900   

   gdp_per_capita ($)       generation  
0               796.0     Generation X  
1               796.0           Silent  
2               796.0     Generation X  
3               796.0  G.I. Generation  
4               796.0          Boomers

In [28]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11009 entries, 0 to 11008
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   country             11009 non-null  object 
 1   year                11008 non-null  float64
 2   sex                 11008 non-null  object 
 3   age                 11008 non-null  object 
 4   suicides_no         11008 non-null  float64
 5   population          11008 non-null  float64
 6   suicides/100k pop   11008 non-null  float64
 7   country-year        11008 non-null  object 
 8   HDI for year        3340 non-null   float64
 9    gdp_for_year ($)   11008 non-null  object 
 10  gdp_per_capita ($)  11008 non-null  float64
 11  generation          11008 non-null  object 
dtypes: float64(6), object(6)
memory usage: 1.0+ MB


In [16]:
print(df.isna().sum())

country                  0
year                     1
sex                      1
age                      1
suicides_no              1
population               1
suicides/100k pop        1
country-year             1
HDI for year          7669
 gdp_for_year ($)        1
gdp_per_capita ($)       1
generation               1
dtype: int64


In [17]:
print(df.duplicated().sum())

0


In [25]:
data = pd.DataFrame({'country': df['country'], 'year': df['year'], 'sex': df['sex'], 'age': df['age'], 'suicides_no': df['suicides_no'], 'population': df['population'], 'suicides': df['suicides/100k pop'], 'country_year': df['country-year'], 'hdi': df['HDI for year'], 'gdp_for_year': df[' gdp_for_year ($) '], 'gdp_per_capita': df['gdp_per_capita ($)'], 'generation': df['generation'], })
data.head(5)

country    year     sex          age  suicides_no  population  suicides  \
0  Albania  1987.0    male  15-24 years         21.0    312900.0      6.71   
1  Albania  1987.0    male  35-54 years         16.0    308000.0      5.19   
2  Albania  1987.0  female  15-24 years         14.0    289700.0      4.83   
3  Albania  1987.0    male    75+ years          1.0     21800.0      4.59   
4  Albania  1987.0    male  25-34 years          9.0    274300.0      3.28   

  country_year  hdi   gdp_for_year  gdp_per_capita       generation  
0  Albania1987  NaN  2,156,624,900           796.0     Generation X  
1  Albania1987  NaN  2,156,624,900           796.0           Silent  
2  Albania1987  NaN  2,156,624,900           796.0     Generation X  
3  Albania1987  NaN  2,156,624,900           796.0  G.I. Generation  
4  Albania1987  NaN  2,156,624,900           796.0          Boomers

In [26]:
#Убедитесь в том, что данные прошли предподготовку. 
#Пропущенные и неопределённые значения выявляет метод isna(), а суммарное количество таких значений — метод sum().
data['hdi']=data['hdi'].fillna(0)
data.dropna(inplace=True)
print(data.isna().sum())

country           0
year              0
sex               0
age               0
suicides_no       0
population        0
suicides          0
country_year      0
hdi               0
gdp_for_year      0
gdp_per_capita    0
generation        0
dtype: int64


In [30]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 11008 entries, 0 to 11007
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   country         11008 non-null  object 
 1   year            11008 non-null  float64
 2   sex             11008 non-null  object 
 3   age             11008 non-null  object 
 4   suicides_no     11008 non-null  float64
 5   population      11008 non-null  float64
 6   suicides        11008 non-null  float64
 7   country_year    11008 non-null  object 
 8   hdi             11008 non-null  float64
 9   gdp_for_year    11008 non-null  object 
 10  gdp_per_capita  11008 non-null  float64
 11  generation      11008 non-null  object 
dtypes: float64(6), object(6)
memory usage: 1.1+ MB


In [36]:
#Сгруппируйте DataFrame по столбцу country, сохраните полученный результат в переменной country_grouping.
country_grouping = data.groupby('country')

In [60]:
#Посчитайте количество случаев самоубийств по странам, которые были совершены, методом sum(), 
#указав, что выбираем один столбец suicides_no.
#Сохраните результат в переменной country_sum и выведите первые 30 строк этой таблицы.
country_sum = country_grouping['suicides_no'].sum().reset_index(name='sum')
country_sum.head(30)

country       sum
0                  Albania    1970.0
1      Antigua and Barbuda      11.0
2                Argentina   82219.0
3                  Armenia    1905.0
4                    Aruba     101.0
5                Australia   70111.0
6                  Austria   50073.0
7               Azerbaijan    1656.0
8                  Bahamas      93.0
9                  Bahrain     463.0
10                Barbados     177.0
11                 Belarus   59892.0
12                 Belgium   62761.0
13                  Belize     348.0
14  Bosnia and Herzegovina     318.0
15                  Brazil  226613.0
16                Bulgaria   36388.0
17              Cabo Verde      42.0
18                  Canada  107561.0
19                   Chile   40895.0
20                Colombia   53080.0
21              Costa Rica    6792.0
22                 Croatia   18429.0
23                    Cuba   41418.0
24                  Cyprus     412.0
25          Czech Republic   43687.0
26                 Denmark   15297.0
27                Dominica       0.0
28                 Ecuador   20660.0
29             El Salvador   11683.0

# Сортировка данных


Поиск необычного в группе — что среди планет, что среди меломанов — это прежде всего поиск чемпионов: объектов с выдающимися показателями по разным статьям. Как всю таблицу, так и отдельные группы изучают, сортируя строки по какому-либо столбцу. 

В Pandas для этой операции есть метод **sort_values()**. У него два аргумента:

• **by = 'имя столбца'** — имя столбца, по которому нужно сортировать;

• **ascending:** по умолчанию True. Для сортировки по убыванию установите значение False.

![image.png](attachment:image.png)


Среди экзопланет интересны близкие по размерам к Земле. Есть ли такие? Отсортируем список по радиусу в порядке возрастания. Тогда в голове таблицы окажутся самые малые, на которых гравитация не прижмёт нас к полу.

**print(exoplanet.sort_values(by='radius').head(30))**

Оказывается, некоторые из уже открытых экзопланет по размерам близки не то что к Земле, но уже и к Луне! Получим список экзопланет с радиусом меньше земного. Смотрите, как логический оператор (здесь это <) задействуется в отборе элементов столбца. Дальше нам этот приём не раз понадобится.

**print(exoplanet[exoplanet['radius'] < 1])**

Но и этот список такой длинный, что изучать его лучше по частям. Экзопланеты, близкие по размерам к Земле, за последнее десятилетие открывали нередко. Можно изучать список открытых за каждый год. Например, для 2014 года (вновь обратите внимание на работу логического оператора, теперь это ==):

**print(exoplanet[exoplanet['discovered'] == 2014])**

А чтобы не тратить время на лишнее, поставим оба условия сразу. Для этого в Pandas есть логический оператор &, подобный оператору and языка Python. Напомним, его смысл на русском языке можно передать словами «и ещё»:

**exo_small_14 = exoplanet[ (exoplanet['radius']<1) & (exoplanet['discovered']==2014)]**

**print(exo_small_14)**

Отсортируем результат в порядке убывания радиуса.

**print(exo_small_14.sort_values(by = 'radius', ascending = False))**





## Задача
1. Выделим пятерки стран с самым низким и самым высоким уровнем самоубийств в мире.

In [62]:
# C самым низким уровнем самоубийств в мире
country_sum.sort_values(by = 'sum', ascending = True).head(5)

country   sum
27             Dominica   0.0
1   Antigua and Barbuda  11.0
37              Grenada  38.0
17           Cabo Verde  42.0
8               Bahamas  93.0

In [64]:
# C самым высоким уровнем самоубийств в мире
country_sum.sort_values(by = 'sum', ascending = False).head(5)


country       sum
33     France  329127.0
35    Germany  291262.0
15     Brazil  226613.0
18     Canada  107561.0
2   Argentina   82219.0

2. Исследуйте распределение количества суидидов по гендерному признаку. Найдите лидеров и антилидеров по количеству самоубийств среди женщин. Повторяется ли картина по сравнению с предыдущим исследованием?

In [70]:
sex_grouping = data.groupby(['country', 'sex'])
sex_grouping.head(1)


country    year     sex          age  suicides_no  \
0                  Albania  1987.0    male  15-24 years         21.0   
2                  Albania  1987.0  female  15-24 years         14.0   
264    Antigua and Barbuda  1985.0  female  15-24 years          0.0   
270    Antigua and Barbuda  1985.0    male  15-24 years          0.0   
588              Argentina  1985.0    male    75+ years        202.0   
...                    ...     ...     ...          ...          ...   
10026               Greece  1985.0  female    75+ years         15.0   
10394              Grenada  1985.0    male  55-74 years          1.0   
10395              Grenada  1985.0  female  15-24 years          0.0   
10704            Guatemala  1986.0    male  25-34 years         46.0   
10709            Guatemala  1986.0  female  15-24 years         17.0   

       population  suicides             country_year    hdi    gdp_for_year  \
0        312900.0      6.71              Albania1987  0.000   2,156,624,900   
2        289700.0      4.83              Albania1987  0.000   2,156,624,900   
264        7709.0      0.00  Antigua and Barbuda1985  0.000     240,923,926   
270        7667.0      0.00  Antigua and Barbuda1985  0.000     240,923,926   
588      363000.0     55.65            Argentina1985  0.694  88,416,668,900   
...           ...       ...                      ...    ...             ...   
10026    317900.0      4.72               Greece1985  0.741  47,820,850,975   
10394      3987.0     25.08              Grenada1985  0.000     167,728,444   
10395     11725.0      0.00              Grenada1985  0.000     167,728,444   
10704    522775.0      8.80            Guatemala1986  0.000   7,231,963,516   
10709    807507.0      2.11            Guatemala1986  0.000   7,231,963,516   

       gdp_per_capita       generation  
0               796.0     Generation X  
2               796.0     Generation X  
264            3850.0     Generation X  
270            3850.0     Generation X  
588            3264.0  G.I. Generation  
...               ...              ...  
10026          5156.0  G.I. Generation  
10394          1965.0  G.I. Generation  
10395          1965.0     Generation X  
10704          1044.0          Boomers  
10709          1044.0     Generation X  

[78 rows x 12 columns]

In [73]:
sex_sum = sex_grouping['suicides_no'].sum().reset_index(name='sum')
female_sum = sex_sum[sex_sum['sex'] == 'female']
# C самым низким уровнем самоубийств в мире среди женщин
female_sum.sort_values(by = 'sum', ascending = True).head(5)

country     sex   sum
54             Dominica  female   0.0
2   Antigua and Barbuda  female   1.0
74              Grenada  female   4.0
34           Cabo Verde  female   5.0
16              Bahamas  female  14.0

In [74]:
# C самым высоким уровнем самоубийств в мире среди женщин
female_sum.sort_values(by = 'sum', ascending = False).head(5)

country     sex      sum
66     France  female  89419.0
70    Germany  female  79707.0
30     Brazil  female  49015.0
36     Canada  female  24184.0
4   Argentina  female  18326.0

# Описательная статистика

Чтобы представить себе среднестатистического самоубийцу, понадобится описательная статистика.

Из её показателей для количественного описания данных в нашей задаче нужны четыре меры — максимум, минимум, медиана и среднее.

Наибольшее и наименьшее обычно вычисляют только по одному признаку. Например, можно получить минимальное и максимальное значение уровня самоубийств (количество самоубийств на 100 тыс.населения) (столбец 'suicides/100k pop'). Для поиска максимума вызывают метод max(), примененный к этому столбцу:

print(master['suicides/100k pop'].max())


In [75]:
print(data['suicides'].max())

224.97


Самый высокий показатель оказался 224.97 самоубийств на 100 тыс.населения. Интересно, в какой стране такая ситуация.

Запросим из master строку с максимальным значением, прибегнув к логической индексации с условием

**master['suicides/100k pop'] == master['suicides/100k pop'].max()**

In [77]:
data[data['suicides'] == data['suicides'].max()]

country    year   sex        age  suicides_no  population  suicides  \
1258   Aruba  1995.0  male  75+ years          2.0       889.0    224.97   

     country_year  hdi   gdp_for_year  gdp_per_capita       generation  
1258    Aruba1995  0.0  1,320,670,391         17949.0  G.I. Generation

Минимальное значение — ищут методом **min()**. Нам интересно установить случаи, когда уровень суицидов минимальный, но не нулевой.

Вот как мы создаём выборку  и находим в ней минимальное значение:

**master_drop_null = master[master['suicides/100k pop'] != 0]**

**print(master_drop_null['suicides/100k pop'].min())** 


In [79]:
master_drop_null = data[data['suicides'] != 0]

print(master_drop_null['suicides'].min())

0.06


Получим название самой "счастливой" страны. Как и при поиске максимума, воспользуемся условием и логической индексацией

In [81]:
data[data['suicides'] == master_drop_null['suicides'].min()]['country']

9229        France
10907    Guatemala
Name: country, dtype: object

Результат вывода — две страны, которые соответствуют минимальному значению.
На основе полученных данных можно сделать вывод, что число совершенных самоубийств на 100 тыс. населения  находится в диапазоне от 0,02 до 224.97 секунд, не включая пропущенные.

Это знание пригодится, чтобы разобраться с медианой и средним арифметическим.

Среднее и медиана оценивают значения в центре выборки. Если тех строк, где число с большим количеством самоубийц, столько же, сколько тех, где это число мало — среднее подойдет. Но когда есть оторванные от основной массы лидеры, их результаты сильно смещают значение среднего вверх. Вот почему оценивать в этом случае лучше медианой.

![image.png](attachment:image.png)

В геометрии медиана делит фигуру на две равные по площади. В статистике она делит выборку пополам: в одной половине значения меньше медианного, в другой больше. Логично, что для определения медианы список обязательно должен быть отсортирован — либо по возрастанию, либо по убыванию.
Когда количество значений нечётное, медиана будет равна тому значению, которое оказалось ровно посередине отсортированного набора. Если же количество данных чётное, то медиана рассчитывается как среднее арифметическое двух соседних чисел в середине набора.



В Pandas есть метод **median()**, который считает медиану. По аналогии с **min()** и **max()** его можно применять ко всей таблице, к отдельному столбцу или к сгруппированным данным. Теперь применим медиану ко всем значениям времени прослушивания в нашей таблице, исключив нулевые:



In [86]:
data['suicides'].median()

4.74

Чтобы убедиться, что лидеры действительно смещают средний показатель вверх, найдём среднее арифметическое всех этих значений методом mean():

In [87]:
data['suicides'].mean()

11.226767805232605